### ISSUE TITLE
Ask for 'save document' if dirty documents when quitting.

### ISSUE THREAD

**Issue by boltex** [boltex/leointeg#142]
Ask for 'save document' if dirty Leo documents remain when closing VSCode.

**Comment by edreamleo** [comment]
@boltex This should have been fixed for b1!  I lost some notes before realizing this issue was still open. It absolutely must be fixed before 1.0.

**Comment by boltex** [comment]
Oops! yep, that's perhaps an important one indeed...

**Comment by boltex** [comment]
Hmm, Ive gone through the extension API and seems there's no 'beforeClosing' hook of any kind... and the deactivate call doesn't seem to allow me to show a modal dialog, (I've just tried, and a couple more things to circumvent this etc.) to at least ask to save if Leo documents are still dirty... Turns out there'S only the 'deactivate' method, and it's not blocking - just allows to finish doing stuff without any gui, etc.

😟

![5h4los](https://user-images.githubusercontent.com/847716/126414891-5b2f7cd1-038c-4b9a-963e-35babbbd532c.jpg)

Not too great, but at least we've got a use case to propose to vscode for them to add this possibility :)

**Comment by boltex** [comment]
created https://github.com/microsoft/vscode/issues/129082

**Comment by edreamleo** [comment]
@boltex It's hard to believe that there is no way to do this. If, without leoInteg, the user opens a file, modifies the file, and then attempts to close the file, surely vs-code must ask the user whether to save the file. How could that not be?

**Comment by edreamleo** [comment]
@boltex Update. I just ran the following experiment. In the vs-code explorer, I added a space to the end of one line in commandStack.ts. I then quit vs-code.  No dialog!!

However, when I reopened vs-code, the explorer contains a notification: 
"one unsaved file".

So the explorer *doesn't* save the file externally, but *does* remember that the file is "dirty" the next time the explorer opens the file. Moreover, the explorer remembers the insertion point (right after the inserted blank)!  Naturally, git reports no change to the file.

All this is pretty impressive, but I'm not sure how to replicate the behavior in leoInteg.

**Comment by edreamleo** [comment]
I googled "vs code doesn't warn if file not saved". I found [this issue](https://github.com/microsoft/vscode/issues/21228) which mentions:

You can easily disable this with "files.hotExit": "off".

**Comment by edreamleo** [comment]
@boltex One workaround, perhaps too "heroic", would be to open (in the explorer) "shadow" versions of .leo files and any changed external files. I call this an heroic solution because then leoInteg would have to update (but not save) the shadow files any time a change is made to any file that leoInteg is managing.

I confirmed the following:

1. With the "hot" setting off, any change to an explorer file prompts a dialog when vs code exits.
2. vs-code does *not* synchronize the shadow file with the corresponding file managed by leoInteg. The shadow file can be dirty without affecting the managed file, and vice-versa.

**Comment by edreamleo** [comment]
@boltex Another possible workaround: schedule an "idle-time" save of all dirty managed files. I might still be possible to lose data when vs-code exists, but this workaround would limit the damage.

**Comment by edreamleo** [comment]
@boltex I've removed the "Blocker" label. I do think this issue needs a warning label of some kind.

**Comment by edreamleo** [comment]
@boltex I wonder whether leoserver.py could help. To make this work, leoInteg would have to create the server so that the serer won't automatically die when vs-code exits.

If this is possible, then the server would kill itself only if it got a shutdown request. Otherwise, the server would detect the loss of the socket and would issue a "save" dialog if any files open in Leo's bridge were open. Either way, the server would then die.

Do you think this plan might work?

**Comment by boltex** [comment]
@edreamleo I'll start by making a new setting in leointeg, which will be sent to the server as soon as its starts (like all others settings concerning the server) that will tell the server what to do with unsaved files when the client disconnects (when quitting) It will either be: save all named dirty files, or just ignore all.

I think this will be a good compromise for 1.0 (we cant make the perfect product for 1.0 but i will try to make it bug-free so it doesnt look like a 'beta') Also, having a real product helps in asking vscode's team for a new feature! (such as async-wait for extension's dialogs on quitting)

Status is that so far, although I can detect when quitting, I cant block the quitting, nor open vscode dialogs. 

About your idea/suggestion: I dont know if the python process started by the nodejs process can open dialogs of any kind...

**Comment by edreamleo** [comment]
On Thu, Jul 29, 2021 at 10:03 PM Felix ***@***.***> wrote:

> @edreamleo <https://github.com/edreamleo> I'll start by making a new
> setting in leointeg, which will be sent to the server as soon as its starts
> (like all others settings concerning the server) that will tell the server
> what to do with unsaved files when the client disconnects (when quitting)
> It will either be: save all named dirty files, or just ignore all.
>

I'm glad my idea is worth looking at.

> Status is that so far, although I can detect when quitting, I cant block
> the quitting, nor open vscode dialogs.
>
Good luck!

> About your idea/suggestion: I dont know if the python process started by
> the nodejs process can open dialogs of any kind...
>
I was thinking that the server might open a Tk dialog. Python usually comes
with Tk installed. It's unclear how node.js could possibly *prevent* any
python code from executing. Actually, the same remark goes for a Qt dialog
if, as is likely, the user has also installed Qt.

In short, the main question seems to be whether leoserver.py can be run in
an independent process that won't be killed when vs-code quits.

Edward

**Comment by boltex** [comment]
Indeed, The process is not killed, I'm forcefully killing it because
it stays there indefinitely otherwise! (and we can reconnect to
the leobridge, but i'm putting this for 2.0 with multi user support and
reconnects) so i guess it's worth a try! :)

You might be onto something if it'S true that most people have a gui lib
installed by default...

If you have time, can you shoot me an example of a simple 'ok' dialog with
what is considered the library that is somewhat "installed by default for
sure" with a common python installation on any OS ? (if such a thing even
makes sense?)

Something like this :

*def save_on_quit():*
*    import the_common_lib*
*    the_common_lib.showOkDialog()*

...but with the 'the_common_lib' replaced with something that makes sense.

And from there I'll flesh out the rest (if a simple test like this actually
works as a proof of concept that would be a great solution to this
problem!)
--
Félix

**Comment by boltex** [comment]
(sorry email replies don't support markdown syntax, this is what I tries to paste in the comment above)
Something like this :
```

def save_on_quit():
    import the_common_lib
    the_common_lib.showOkDialog()

```

**Comment by edreamleo** [comment]
On Fri, Jul 30, 2021 at 8:35 PM Felix ***@***.***> wrote:

> Something like this :
>
> def save_on_quit():
>     import the_common_lib
>     the_common_lib.showOkDialog()
>
>
See the g.EmergencyDialog class:  For example, this appears in
app.cleanLeoID:

 g.EmergencyDialog(
    title=f"Invalid Leo ID: {tag}",
    message=(
        f"Invalid Leo ID: {old_id!r}\n\n"
        "Your id should contain only letters and numbers\n"
        "and must be at least 3 characters in length."))

The init code in the class imports Tk, so something like the code above is
all you need for testing.

Imo, however, the server code should try to import Qt using leo.core.leoQt
as usual, and use a Tk dialog only as a last resort. The Tk dialog could
even suggest installing Qt using `pip install` :-)

I'd be happy to add the dialog code to leoserver.py.

Edward

**Comment by edreamleo** [comment]
@boltex Can this issue be closed? I am going to close [Leo issue 2108](https://github.com/leo-editor/leo-editor/issues/2108)
